File: android/app/src/main/java/com/guardforce/guard/ui/Labels.kt
package com.guardforce.guard.ui

import android.content.Context
import com.guardforce.guard.R
import com.guardforce.guard.core.AppError
import com.guardforce.guard.domain.TimeText
import com.guardforce.guard.ui.components.Tone

/** Status → label/tone, mirroring dashboard/src/lib/domain/status.ts. */
object Labels {
    fun attendance(ctx: Context, a: String): Pair<String, Tone> = when (a) {
        "present" -> ctx.getString(R.string.attendance_present) to Tone.Present
        "half_day" -> ctx.getString(R.string.attendance_half_day) to Tone.HalfDay
        "absent" -> ctx.getString(R.string.attendance_absent) to Tone.Absent
        "on_leave" -> ctx.getString(R.string.attendance_on_leave) to Tone.OnLeave
        else -> ctx.getString(R.string.attendance_pending) to Tone.Neutral
    }

    fun flag(ctx: Context, f: String): String = when (f) {
        "OUTSIDE_FENCE" -> ctx.getString(R.string.flag_outside_fence)
        "LATE_START" -> ctx.getString(R.string.flag_late_start)
        "EARLY_CHECKOUT" -> ctx.getString(R.string.flag_early_checkout)
        "LOCATION_OFF" -> ctx.getString(R.string.flag_location_off)
        "TAMPER_SUSPECTED" -> ctx.getString(R.string.flag_tamper)
        "SYNCED_LATE" -> ctx.getString(R.string.flag_synced_late)
        "LOW_ACCURACY" -> ctx.getString(R.string.flag_low_accuracy)
        else -> f
    }

    fun patrol(ctx: Context, s: String): Pair<String, Tone> = when (s) {
        "in_progress" -> ctx.getString(R.string.patrol_status_in_progress) to Tone.Olive
        "completed" -> ctx.getString(R.string.patrol_status_completed) to Tone.Present
        "late" -> ctx.getString(R.string.patrol_status_late) to Tone.HalfDay
        "missed" -> ctx.getString(R.string.patrol_status_missed) to Tone.Absent
        else -> ctx.getString(R.string.patrol_status_scheduled) to Tone.Neutral
    }

    fun leave(ctx: Context, s: String): Pair<String, Tone> = when (s) {
        "approved" -> ctx.getString(R.string.leave_status_approved) to Tone.Present
        "declined" -> ctx.getString(R.string.leave_status_declined) to Tone.Absent
        "cancelled" -> ctx.getString(R.string.leave_status_cancelled) to Tone.Neutral
        else -> ctx.getString(R.string.leave_status_pending) to Tone.HalfDay
    }

    fun leaveType(ctx: Context, t: String): String = when (t) {
        "earned" -> ctx.getString(R.string.leave_earned)
        "unpaid" -> ctx.getString(R.string.leave_unpaid)
        else -> ctx.getString(R.string.leave_casual)
    }

    fun docType(ctx: Context, t: String): String = when (t) {
        "aadhaar" -> ctx.getString(R.string.doc_aadhaar); "pan" -> ctx.getString(R.string.doc_pan)
        "police_verification" -> ctx.getString(R.string.doc_police_verification); "marksheet" -> ctx.getString(R.string.doc_marksheet)
        "guard_kyc" -> ctx.getString(R.string.doc_guard_kyc); else -> ctx.getString(R.string.doc_other)
    }

    fun docStatus(ctx: Context, s: String?): Pair<String, Tone> = when (s) {
        "verified" -> ctx.getString(R.string.doc_verified) to Tone.Present
        "rejected" -> ctx.getString(R.string.doc_rejected) to Tone.Absent
        "pending" -> ctx.getString(R.string.doc_pending) to Tone.HalfDay
        else -> ctx.getString(R.string.doc_missing) to Tone.Neutral
    }

    fun kycGap(ctx: Context, g: String): String = when (g) {
        "phone_verification" -> ctx.getString(R.string.gap_phone_verification)
        "registration_selfie" -> ctx.getString(R.string.gap_registration_selfie)
        "designation" -> ctx.getString(R.string.gap_designation)
        else -> docType(ctx, g)
    }

    fun relative(ctx: Context, r: TimeText.Relative): String = when (r) {
        TimeText.Relative.Now -> ctx.getString(R.string.time_now)
        is TimeText.Relative.Minutes -> ctx.getString(R.string.time_min_ago, r.n)
        is TimeText.Relative.Hours -> ctx.getString(R.string.time_hr_ago, r.n)
    }

    fun duration(ctx: Context, seconds: Long): String {
        val h = seconds / 3600; val m = (seconds % 3600) / 60
        return if (h > 0) ctx.getString(R.string.duration_hm, h.toInt(), m.toInt()) else ctx.getString(R.string.duration_m, m.toInt())
    }

    fun error(ctx: Context, e: Throwable): String = when (e) {
        is AppError.Network -> ctx.getString(R.string.error_network)
        is AppError.Unauthorized -> ctx.getString(R.string.error_session)
        is AppError.Rpc -> when (e.code) {
            "NO_GUARD_FOR_PHONE" -> ctx.getString(R.string.claim_not_found)
            "PHONE_ALREADY_LINKED" -> ctx.getString(R.string.claim_already_linked)
            "TAMPER_SUSPECTED" -> ctx.getString(R.string.checkin_mock)
            "LEAVE_OVERLAP" -> ctx.getString(R.string.leave_overlap)
            "LEAVE_DATES_INVALID" -> ctx.getString(R.string.leave_dates_invalid)
            "PATROL_PHOTO_REQUIRED" -> ctx.getString(R.string.patrol_photo_required, 1)
            "TASK_PHOTO_REQUIRED" -> ctx.getString(R.string.task_photo_required)
            else -> ctx.getString(R.string.error_generic, e.code)
        }
        is AppError.Http -> if (e.status in 400..499 && e.detail.contains("otp", true) || e.detail.contains("token", true)) ctx.getString(R.string.otp_invalid) else ctx.getString(R.string.error_generic, e.detail.take(80))
        else -> ctx.getString(R.string.error_generic, e.message?.take(80) ?: "?")
    }
}

File: guard-app/src/ui/labels.ts
import { ApiError } from "../api/errors";
import { t, type StringKey } from "../i18n";
import type { Tone } from "./theme";

/** Status → label/tone, mirroring dashboard/src/lib/domain/status.ts. */
export const attendance = (a: string): [string, Tone] =>
  a === "present" ? [t("attendance_present"), "present"] : a === "half_day" ? [t("attendance_half_day"), "halfDay"] : a === "absent" ? [t("attendance_absent"), "absent"] : a === "on_leave" ? [t("attendance_on_leave"), "onLeave"] : [t("attendance_pending"), "neutral"];
const flagKeys: Record<string, StringKey> = { OUTSIDE_FENCE: "flag_outside_fence", LATE_START: "flag_late_start", EARLY_CHECKOUT: "flag_early_checkout", LOCATION_OFF: "flag_location_off", TAMPER_SUSPECTED: "flag_tamper", SYNCED_LATE: "flag_synced_late", LOW_ACCURACY: "flag_low_accuracy" };
export const flag = (f: string) => (flagKeys[f] ? t(flagKeys[f]) : f);
export const flagTone = (f: string): Tone => (f === "LOCATION_OFF" || f === "TAMPER_SUSPECTED" ? "absent" : "halfDay");
export const patrol = (s: string): [string, Tone] =>
  s === "in_progress" ? [t("patrol_status_in_progress"), "olive"] : s === "completed" ? [t("patrol_status_completed"), "present"] : s === "late" ? [t("patrol_status_late"), "halfDay"] : s === "missed" ? [t("patrol_status_missed"), "absent"] : [t("patrol_status_scheduled"), "neutral"];
export const leave = (s: string): [string, Tone] =>
  s === "approved" ? [t("leave_status_approved"), "present"] : s === "declined" ? [t("leave_status_declined"), "absent"] : s === "cancelled" ? [t("leave_status_cancelled"), "neutral"] : [t("leave_status_pending"), "halfDay"];
export const leaveType = (ty: string) => (ty === "earned" ? t("leave_earned") : ty === "unpaid" ? t("leave_unpaid") : t("leave_casual"));
const docKeys: Record<string, StringKey> = { aadhaar: "doc_aadhaar", pan: "doc_pan", police_verification: "doc_police_verification", marksheet: "doc_marksheet", guard_kyc: "doc_guard_kyc" };
export const docType = (d: string) => t(docKeys[d] ?? "doc_other");
export const docStatus = (s: string | null | undefined): [string, Tone] =>
  s === "verified" ? [t("doc_verified"), "present"] : s === "rejected" ? [t("doc_rejected"), "absent"] : s === "pending" ? [t("doc_pending"), "halfDay"] : [t("doc_missing"), "neutral"];
export const kycGap = (g: string) => (g === "phone_verification" ? t("gap_phone_verification") : g === "registration_selfie" ? t("gap_registration_selfie") : g === "designation" ? t("gap_designation") : docType(g));
export const relative = (r: { unit: "now" | "min" | "hr"; n: number }) => (r.unit === "now" ? t("time_now") : r.unit === "min" ? t("time_min_ago", r.n) : t("time_hr_ago", r.n));
export const duration = (seconds: number) => { const h = Math.floor(seconds / 3600), m = Math.floor((seconds % 3600) / 60); return h > 0 ? t("duration_hm", h, String(m).padStart(2, "0")) : t("duration_m", m); };

export function errorText(e: unknown): string {
  const err = e instanceof ApiError ? e : null;
  const code = err?.code ?? "";
  switch (code) {
    case "NETWORK": case "SERVER": return t("error_network");
    case "UNAUTHORIZED": return t("error_session");
    case "NO_GUARD_FOR_PHONE": return t("claim_not_found");
    case "PHONE_ALREADY_LINKED": return t("claim_already_linked");
    case "TAMPER_SUSPECTED": return t("checkin_mock");
    case "LEAVE_OVERLAP": return t("leave_overlap");
    case "LEAVE_DATES_INVALID": return t("leave_dates_invalid");
    case "PATROL_PHOTO_REQUIRED": return t("patrol_photo_required", 1);
    case "TASK_PHOTO_REQUIRED": return t("task_photo_required");
  }
  const msg = (e as Error)?.message ?? String(e);
  if (/otp|token/i.test(msg)) return t("otp_invalid");
  return t("error_generic", (code || msg).slice(0, 80));
}
